CONTRIBUTING.md

```diff
@@ -205,6 +205,209 @@ Please follow the coding style of the project. Material-UI uses prettier and esl
 
 Finally, when you submit a Pull Request, they are run again by our continuous integration tools, but hopefully, your code is already clean!
 
+## Convention Typescript Component
+
+follow this convention if the component is in `.tsx` (both core and lab components)
+
+### Public Component
+
+refer to components that are exported from `@material-ui/core` or `@material-ui/lab`.
+
+```tsx
+import { SxProps } from '@material-ui/system';
+import {
+  experimentalStyled as styled,
+  unstable_useThemeProps as useThemeProps,
+  Theme,
+} from '@material-ui/core/styles';
+
+// 1st: props interface of the component with `classes`, `sx` and other props in alphabetical order
+export interface FooProps {
+  // add comment to each key for api docs
+  /**
+   * Override or extend the styles applied to the component.
+   */
+  classes?: {
+    /** Styles applied to the root element. */
+    root?: string;
+    /** Styles applied to the root element if `disableMargin=false`. */
+    bar?: string;
+    /** Pseudo-class applied to the root element if `disabled=true`. */
+    disabled?: string;
+  };
+  ...props
+  /**
+   * The system prop that allows defining system overrides as well as additional CSS styles.
+   */
+  sx?: SxProps<Theme>;
+}
+// in case that it extends from public component
+// export interface FooProps extends TypographyProps
+
+// 2nd: ClassKey
+export type FooClassKey = keyof NonNullable<FooProps['classes']>;
+// in case that props interface require generic argument, pass `unknown`
+// Ex. export type FooClassKey = keyof NonNullable<FooProps<unknown>['classes']>;
+
+// 3rd: classes generator & utility
+export function getFooUtilityClass(slot: string) {
+  return generateUtilityClass('MuiFooDay', slot);
+}
+// make sure it has ClassKey as arg in generic
+export const fooClasses = generateUtilityClasses<FooClassKey>('MuiFoo', [
+  'root',
+  'bar',
+  'disabled',
+]);
+
+const useUtilityClasses = (styleProps: FooProps & { extraProp: boolean}) => {
+  // extraProp might be the key/value from react context that this component access
+  const {
+    bar,
+    disabled,
+    classes,
+  } = styleProps;
+
+  const slots = {
+    root: [
+      'root',
+      bar && 'bar',
+      disabled && 'disabled',
+    ],
+  };
+
+  return composeClasses(slots, getFooUtilityClass, classes);
+};
+
+// 4th: StyledComponent, similar to JS convention
+// add additional props in between to append props interface to StyledComponent
+const FooRoot = styled(
+  Typography,
+  {},
+  {
+    name: 'MuiFoo',
+    slot: 'Root',
+    overridesResolver: (props, styles) => styles.root,
+  },
+)<{ component?: React.ElementType }>(...);
```

Comment:
Should we add here `styleProps` too? Make them required & make the types stricter?